File: backend/internal/service/badges.go
package service

import (
	"context"
	"fmt"
	"strings"
	"time"

	"github.com/jackc/pgx/v5/pgtype"
	"github.com/nishantg96/gitfable/internal/database"
)

// MergedDraw is a flattened representation of a merged draw with its issue data,
// used for badge condition checking without depending on sqlc types.
type MergedDraw struct {
	ID        int64
	CreatedAt time.Time
	MergedAt  time.Time
	RepoOwner string
	RepoName  string
	Language  string
	Labels    []string
}

// BadgeUser holds user-level fields needed by badge checks.
type BadgeUser struct {
	LongestStreak int32
}

// BadgeDef describes a badge and its unlock condition.
type BadgeDef struct {
	Name        string
	Description string
	Icon        string
	Check       func(draws []MergedDraw, user BadgeUser) bool
}

// BadgeDefs is the canonical list of all 10 badges.
var BadgeDefs = []BadgeDef{
	{
		Name:        "Prologue",
		Description: "Merge your first pull request",
		Icon:        "book-open",
		Check: func(draws []MergedDraw, _ BadgeUser) bool {
			return len(draws) >= 1
		},
	},
	{
		Name:        "Short Story",
		Description: "Merge 3 pull requests",
		Icon:        "book",
		Check: func(draws []MergedDraw, _ BadgeUser) bool {
			return len(draws) >= 3
		},
	},
	{
		Name:        "The Epic",
		Description: "Merge 100 pull requests",
		Icon:        "library",
		Check: func(draws []MergedDraw, _ BadgeUser) bool {
			return len(draws) >= 100
		},
	},
	{
		Name:        "Anthology",
		Description: "Merge pull requests in 5 or more languages",
		Icon:        "globe",
		Check: func(draws []MergedDraw, _ BadgeUser) bool {
			langs := make(map[string]struct{})
			for _, d := range draws {
				if d.Language != "" {
					langs[d.Language] = struct{}{}
				}
			}
			return len(langs) >= 5
		},
	},
	{
		Name:        "Midnight Draft",
		Description: "Merge a pull request between midnight and 5 AM",
		Icon:        "moon",
		Check: func(draws []MergedDraw, _ BadgeUser) bool {
			for _, d := range draws {
				h := d.MergedAt.Hour()
				if h >= 0 && h < 5 {
					return true
				}
			}
			return false
		},
	},
	{
		Name:        "Fast Forward",
		Description: "Merge a pull request within 24 hours of drawing it",
		Icon:        "zap",
		Check: func(draws []MergedDraw, _ BadgeUser) bool {
			for _, d := range draws {
				if d.MergedAt.Sub(d.CreatedAt) < 24*time.Hour {
					return true
				}
			}
			return false
		},
	},
	{
		Name:        "Daily Author",
		Description: "Maintain a 30-day contribution streak",
		Icon:        "flame",
		Check: func(_ []MergedDraw, user BadgeUser) bool {
			return user.LongestStreak >= 30
		},
	},
	{
		Name:        "Worldbuilder",
		Description: "Contribute to 10 or more different repositories",
		Icon:        "map",
		Check: func(draws []MergedDraw, _ BadgeUser) bool {
			repos := make(map[string]struct{})
			for _, d := range draws {
				key := d.RepoOwner + "/" + d.RepoName
				repos[key] = struct{}{}
			}
			return len(repos) >= 10
		},
	},
	{
		Name:        "Proofreader",
		Description: "Merge 10 pull requests with a bug label",
		Icon:        "bug",
		Check: func(draws []MergedDraw, _ BadgeUser) bool {
			count := 0
			for _, d := range draws {
				if hasLabelContaining(d.Labels, "bug") {
					count++
				}
			}
			return count >= 10
		},
	},
	{
		Name:        "The Archivist",
		Description: "Merge 10 pull requests with a documentation label",
		Icon:        "file-text",
		Check: func(draws []MergedDraw, _ BadgeUser) bool {
			count := 0
			for _, d := range draws {
				if hasLabelContaining(d.Labels, "doc") {
					count++
				}
			}
			return count >= 10
		},
	},
}

func hasLabelContaining(labels []string, substr string) bool {
	for _, l := range labels {
		if strings.Contains(strings.ToLower(l), substr) {
			return true
		}
	}
	return false
}

// BadgeStore is the interface for badge-related database operations.
// Method signatures match the sqlc-generated Queries methods exactly.
type BadgeStore interface {
	UpsertBadge(ctx context.Context, arg database.UpsertBadgeParams) error
	GetBadgeByName(ctx context.Context, name string) (database.Badge, error)
	AwardBadge(ctx context.Context, arg database.AwardBadgeParams) (database.UserBadge, error)
	GetUserBadgeNames(ctx context.Context, userID int64) ([]string, error)
	GetUserMergedDrawsWithIssues(ctx context.Context, userID int64) ([]database.GetUserMergedDrawsWithIssuesRow, error)
}

// BadgeService handles badge initialization and awarding.
type BadgeService struct {
	store BadgeStore
}

// NewBadgeService creates a new BadgeService.
func NewBadgeService(store BadgeStore) *BadgeService {
	return &BadgeService{store: store}
}

// InitializeBadges upserts all badge definitions into the database.
func (s *BadgeService) InitializeBadges(ctx context.Context) error {
	for _, bd := range BadgeDefs {
		err := s.store.UpsertBadge(ctx, database.UpsertBadgeParams{
			Name:        bd.Name,
			Description: bd.Description,
			Icon:        bd.Icon,
		})
		if err != nil {
			return fmt.Errorf("upsert badge %q: %w", bd.Name, err)
		}
	}
	return nil
}

// CheckBadges evaluates all badge conditions for a user and awards any newly earned badges.
// It returns the names of newly awarded badges.
func (s *BadgeService) CheckBadges(ctx context.Context, userID int64, user BadgeUser) ([]string, error) {
	// Get existing badge names
	existing, err := s.store.GetUserBadgeNames(ctx, userID)
	if err != nil {
		return nil, fmt.Errorf("get user badge names: %w", err)
	}
	owned := make(map[string]struct{}, len(existing))
	for _, name := range existing {
		owned[name] = struct{}{}
	}

	// Load merged draws with issue data
	rows, err := s.store.GetUserMergedDrawsWithIssues(ctx, userID)
	if err != nil {
		return nil, fmt.Errorf("get merged draws: %w", err)
	}

	draws := make([]MergedDraw, 0, len(rows))
	for _, r := range rows {
		d := MergedDraw{
			ID:        r.ID,
			RepoOwner: r.RepoOwner,
			RepoName:  r.RepoName,
			Labels:    r.Labels,
		}
		if r.CreatedAt.Valid {
			d.CreatedAt = r.CreatedAt.Time
		}
		if r.MergedAt.Valid {
			d.MergedAt = r.MergedAt.Time
		}
		if r.Language.Valid {
			d.Language = r.Language.String
		}
		draws = append(draws, d)
	}

	var awarded []string
	for _, bd := range BadgeDefs {
		if _, already := owned[bd.Name]; already {
			continue
		}
		if !bd.Check(draws, user) {
			continue
		}

		// Look up badge ID
		badge, err := s.store.GetBadgeByName(ctx, bd.Name)
		if err != nil {
			return awarded, fmt.Errorf("get badge %q: %w", bd.Name, err)
		}

		_, err = s.store.AwardBadge(ctx, database.AwardBadgeParams{
			UserID:  userID,
			BadgeID: badge.ID,
			DrawID:  pgtype.Int8{Valid: false},
		})
		if err != nil {
			return awarded, fmt.Errorf("award badge %q: %w", bd.Name, err)
		}
		awarded = append(awarded, bd.Name)
	}

	return awarded, nil
}

File: backend/app/services/badges.py
from datetime import datetime, timezone
from typing import List

from sqlalchemy import select, func, and_, or_
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from app.models.database import Badge, UserBadge, Draw, User, Issue


# Badge definitions (should match database)
BADGES_META = [
    {"name": "Prologue", "description": "First merged PR", "icon": "book-open"},
    {"name": "Short Story", "description": "3 PRs merged", "icon": "files"},
    {"name": "The Epic", "description": "100 PRs merged", "icon": "library"},
    {"name": "Anthology", "description": "PRs in 5+ languages", "icon": "book-copy"},
    {
        "name": "Midnight Draft",
        "description": "PR merged midnight-5am",
        "icon": "moon-star",
    },
    {
        "name": "Fast Forward",
        "description": "PR merged within 24h",
        "icon": "fast-forward",
    },
    {"name": "Daily Author", "description": "30-day streak", "icon": "pen-tool"},
    {"name": "Worldbuilder", "description": "10+ different repos", "icon": "globe"},
    {"name": "Proofreader", "description": "10 bug PRs merged", "icon": "search"},
    {"name": "The Archivist", "description": "10 docs PRs merged", "icon": "bookmark"},
]


async def check_badges(user_id: str, db: AsyncSession) -> List[dict]:
    """Check and award badges to user.

    Args:
        user_id: User UUID string
        db: Database session

    Returns:
        List of newly awarded badges
    """
    from uuid import UUID

    # Get user
    result = await db.execute(select(User).where(User.id == UUID(user_id)))
    user = result.scalar_one_or_none()

    if not user:
        return []

    # Get already earned badges
    result = await db.execute(
        select(Badge.name).join(UserBadge).where(UserBadge.user_id == UUID(user_id))
    )
    earned = {row[0] for row in result.fetchall()}

    # Get merged draws with issue data
    result = await db.execute(
        select(Draw)
        .options(selectinload(Draw.issue))
        .where(Draw.user_id == UUID(user_id))
        .where(Draw.status == "merged")
    )
    merged_draws = result.scalars().all()

    count = len(merged_draws)
    new_badges = []
    now = datetime.now(timezone.utc)

    # Check each badge condition

    # Prologue: First merged PR
    if "Prologue" not in earned and count >= 1:
        new_badges.append(await _award_badge(user_id, "Prologue", db, now))

    # Short Story: 3 PRs merged
    if "Short Story" not in earned and count >= 3:
        new_badges.append(await _award_badge(user_id, "Short Story", db, now))

    # The Epic: 100 PRs merged
    if "The Epic" not in earned and count >= 100:
        new_badges.append(await _award_badge(user_id, "The Epic", db, now))

    # Anthology: PRs in 5+ languages
    if "Anthology" not in earned and count > 0:
        languages = {
            draw.issue.language
            for draw in merged_draws
            if draw.issue and draw.issue.language
        }
        if len(languages) >= 5:
            new_badges.append(await _award_badge(user_id, "Anthology", db, now))

    # Midnight Draft: PR merged midnight-5am
    if "Midnight Draft" not in earned and count > 0:
        for draw in merged_draws:
            if draw.merged_at and draw.merged_at.hour < 5:
                new_badges.append(
                    await _award_badge(user_id, "Midnight Draft", db, now)
                )
                break

    # Fast Forward: PR merged within 24h
    if "Fast Forward" not in earned and count > 0:
        for draw in merged_draws:
            if draw.merged_at and draw.drawn_at:
                diff = (draw.merged_at - draw.drawn_at).total_seconds()
                if diff < 86400:  # 24 hours
                    new_badges.append(
                        await _award_badge(user_id, "Fast Forward", db, now)
                    )
                    break

    # Daily Author: 30-day streak
    if "Daily Author" not in earned and user.longest_streak >= 30:
        new_badges.append(await _award_badge(user_id, "Daily Author", db, now))

    # Worldbuilder: 10+ different repos
    if "Worldbuilder" not in earned and count > 0:
        repos = {draw.issue.repo for draw in merged_draws if draw.issue}
        if len(repos) >= 10:
            new_badges.append(await _award_badge(user_id, "Worldbuilder", db, now))

    # Proofreader: 10 bug PRs merged
    if "Proofreader" not in earned and count > 0:
        bug_count = 0
        for draw in merged_draws:
            if draw.issue and draw.issue.labels:
                if any("bug" in label.lower() for label in draw.issue.labels):
                    bug_count += 1
        if bug_count >= 10:
            new_badges.append(await _award_badge(user_id, "Proofreader", db, now))

    # The Archivist: 10 docs PRs merged
    if "The Archivist" not in earned and count > 0:
        doc_count = 0
        for draw in merged_draws:
            if draw.issue and draw.issue.labels:
                if any("doc" in label.lower() for label in draw.issue.labels):
                    doc_count += 1
        if doc_count >= 10:
            new_badges.append(await _award_badge(user_id, "The Archivist", db, now))

    # Filter out None values (if badge already existed)
    return [b for b in new_badges if b]


async def _award_badge(
    user_id: str, badge_name: str, db: AsyncSession, earned_at: datetime
) -> dict:
    """Award a badge to user.

    Args:
        user_id: User UUID string
        badge_name: Badge name
        db: Database session
        earned_at: Timestamp

    Returns:
        Badge info dict or None if already has badge
    """
    from uuid import UUID

    # Get badge ID
    result = await db.execute(select(Badge).where(Badge.name == badge_name))
    badge = result.scalar_one_or_none()

    if not badge:
        # Badge doesn't exist in database, skip
        return None

    # Check if user already has this badge
    result = await db.execute(
        select(UserBadge)
        .where(UserBadge.user_id == UUID(user_id))
        .where(UserBadge.badge_id == badge.id)
    )
    if result.scalar_one_or_none():
        return None

    # Award badge
    user_badge = UserBadge(
        user_id=UUID(user_id), badge_id=badge.id, earned_at=earned_at
    )
    db.add(user_badge)
    await db.commit()

    return {"name": badge.name, "earned_at": earned_at.isoformat()}


async def get_user_badges(user_id: str, db: AsyncSession) -> List[dict]:
    """Get all badges earned by user.

    Args:
        user_id: User UUID string
        db: Database session

    Returns:
        List of badge dicts with earned_at
    """
    from uuid import UUID

    result = await db.execute(
        select(Badge, UserBadge.earned_at)
        .join(UserBadge)
        .where(UserBadge.user_id == UUID(user_id))
        .order_by(UserBadge.earned_at.desc())
    )

    badges = []
    for badge, earned_at in result.fetchall():
        badges.append(
            {
                "name": badge.name,
                "description": badge.description,
                "icon": badge.icon,
                "earned_at": earned_at.isoformat() if earned_at else None,
            }
        )

    return badges


async def initialize_badges(db: AsyncSession) -> None:
    """Initialize badge definitions in database.

    Should be called on startup to ensure badges exist.
    """
    for badge_meta in BADGES_META:
        # Check if badge exists
        result = await db.execute(select(Badge).where(Badge.name == badge_meta["name"]))
        if not result.scalar_one_or_none():
            # Create badge
            badge = Badge(
                name=badge_meta["name"],
                description=badge_meta["description"],
                icon=badge_meta["icon"],
            )
            db.add(badge)

    await db.commit()
